src/coreclr/src/jit/lower.cpp

```diff
@@ -4498,13 +4504,58 @@ bool Lowering::TryCreateAddrMode(GenTree* addr, bool isContainable)
 // Arguments:
 //    node - the node we care about
 //
-void Lowering::LowerAdd(GenTreeOp* node)
+// Returns:
+//    nullptr if no transformation is done, or the next node in the transformed node sequence that
+//    needs to be lowered.
+//
+GenTree* Lowering::LowerAdd(GenTreeOp* node)
 {
-#ifndef TARGET_ARMARCH
     if (varTypeIsIntegralOrI(node->TypeGet()))
     {
+        GenTree* op1 = node->gtGetOp1();
+        GenTree* op2 = node->gtGetOp2();
+
+        // Check for `value + 0` case, it could survive until lowering when
+        // `0` is `Fseq` with field information that could not be dropped earlier.
+        // We should not access it after lowering, so we can drop it now.
+        GenTree* zero  = nullptr;
+        GenTree* value = nullptr;
+        if (op1->IsCnsIntOrI() && (op1->AsIntCon()->IconValue() == 0))
+        {
+            zero  = op1;
+            value = op2;
+        }
+        else if (op2->IsCnsIntOrI() && (op2->AsIntCon()->IconValue() == 0))
```

Comment:
Typically it's enough to check only `op2` for constants because `gtSetEvalOrder` tries to put constants in the second op.